Theme: Use valid DTCG color format for primitive values

aduth opened the pull request:
## What?

Updates theme package to ensure that primitive design tokens are output in a valid format, per [the DTCG Color Format specification](https://www.designtokens.org/tr/drafts/color/#format).

## Why?

- So that our design tokens are valid DTCG design tokens
- To fix issues with trying to import these tokens in Figma

## How?

We previously had some code to handle this, but it looks like the `if` condition was causing all values not to be matched. The updates here align to expectations that the `color` value passed to the function is a hex value, generated by [`getColorString`](https://github.com/WordPress/gutenberg/blob/c6d4da99618b8675bdc9d8b6bbb01ba9be369c41/packages/theme/src/color-ramps/lib/color-utils.ts#L19-L26).

## Testing Instructions

Verify that running `npm run --workspace @wordpress/theme build` produces no local changes (see `git status`).

Verify that you can import the resulting [`color.json`](https://github.com/WordPress/gutenberg/blob/a6ea63f8568d447bf3c2dabd52100768d10a9eb3/packages/theme/tokens/color.json) into Figma Variables panel, if you have a Figma subscription.

Alternatively, try using an [online validator](https://designtoken-validator.sotec-solutions.com/) to confirm validity. Note that this tool doesn't appear to expect `hex` values, but as referenced in the specification above, `hex` is a valid fallback property.

aduth committed:
Theme: Use valid DTCG color format for primitive values

mirka commented on packages/theme/bin/generate-primitive-tokens/index.ts:
```diff
@@ -30,28 +29,24 @@ const __dirname = path.dirname( __filename );
+		parsed = to( parse( color ), sRGB );
+	} catch {
+		return color;
 	}
```
Not introduced in this PR, but it's not obvious to me what kinds of cases would error on the `parse` but is still safe to silently return with the original value.

aduth replied:
That's a good point, maybe we should avoid catching this, so that it actually disrupts the build for the unexpected input.

mirka reviewed (approved):
I can't check Figma, but the validation is passing 👍 All the color values also seems to be preserved.

aduth committed:
Avoid graceful catching of converted color value

See: https://github.com/WordPress/gutenberg/pull/73858#discussion_r2607066009

The `return` fallback will likely produce an invalid result. We should fail the build if the value can't be parsed.

aduth committed:
Extract function for rounding hex component

Avoid function allocation for each transformed value, since this behavior is static

jsnajdr commented on packages/theme/bin/generate-primitive-tokens/index.ts:
```diff
@@ -29,29 +22,28 @@ const __dirname = path.dirname( __filename );
-	return color;
+	return {
+		colorSpace: 'srgb',
+		components: parsed.coords.map( roundHexComponent ),
```
In version 0.6, when it's released (currently in beta) we'll be able to write `getAll( parsed, { precision: 3 } )`.

Maybe we could switch to 0.6.0-beta2 right away, the stable 0.5.2 version is 17 months old.

aduth replied:
Thanks for the pointer. That'll be nice to have. It sounds like the 0.6.0 release could be imminent, or at least was [planned for "next week"](https://github.com/color-js/color.js/issues/537#issuecomment-3427376480) as of about two months ago 😅

aduth replied:
I think how I'll handle this is set myself a reminder for about a month from now to see if the stable version is released and update both this and your suggestion below. I don't think it's pressing enough to warrant adopting the beta version, unless I'm missing something.

jsnajdr replied:
Yes, it can be updated once a stable version is released, no hurry. I found this suggestion by looking at the `roundHexComponent` code and thinking: wait a minute, many colorjs.io functions already have a `precision` option that does exactly this kind of rounding, can we get the rounded `coords` this way? And the answer turns out to be: yes, in the next version 🙂

jsnajdr commented on packages/theme/bin/generate-primitive-tokens/index.ts:
```diff
@@ -29,29 +22,28 @@ const __dirname = path.dirname( __filename );
-		};
-	}
+const transformColorStringToDTCGValue = ( color: string ) => {
+	const parsed = to( parse( color ), sRGB );
```
Instead of `parse( color )`, we can almost always pass just `color`. Every colorjs.io function calls `getColor` on all color parameters before doing anything else. That's where the parsing will happen.
